drivers: added soft SPI driver

plushvoxel opened the pull request:
This is the initial import for a software SPI driver.
The current version only allow to use either hardware SPI or software SPI.
Right now only writing in master mode is implemented.

The test code sends 0xAA in all four spi modes and then sends "Hello world" in spi mode 0.
![closeup logic analyzer image](https://i.imgur.com/9tZpeOa.png)
![logic analyzer image](https://i.imgur.com/eZP8Yzo.png)

kaspar030 commented:
Nice! This is awesome for testing.

cgundogan commented on drivers/soft_spi/include/soft_spi_params.h:
```diff
@@ -0,0 +1,22 @@
+#endif
+
+soft_spi_t soft_spi_config[] = {
+  #ifdef SOFT_SPI_PARAMS_USER
```
I think user params should be handled like this [1]

[1] https://github.com/RIOT-OS/RIOT/wiki/Guide:-Writing-a-device-driver-in-RIOT#default-device-configuration

aabadie commented on drivers/include/soft_spi.h:
```diff
@@ -0,0 +1,49 @@
+#endif
+
+/**
+ * @brief Sotware SPI port descriptor
```
So**f**tware

aabadie commented on drivers/include/soft_spi.h:
```diff
@@ -0,0 +1,49 @@
+ * @{
+ *
+ * @file
+ * @brief       Sotware SPI port descriptor definition
```
So**f**tware

aabadie commented on drivers/soft_spi/include/soft_spi_params.h:
```diff
@@ -0,0 +1,42 @@
+#endif
+
+/**
+ * @brief   Sotware SPI port descriptor array
```
So**f**tware

aabadie commented on drivers/soft_spi/soft_spi.c:
```diff
@@ -0,0 +1,202 @@
+ * @{
+ *
+ * @file
+ * @brief       Sotware SPI implementation
```
So**f**tware

aabadie commented on drivers/soft_spi/include/soft_spi_params.h:
```diff
@@ -0,0 +1,42 @@
+ * @brief   Sotware SPI port descriptor array
+ */
+soft_spi_t soft_spi_config[] = {
+  //#ifdef SOFT_SPI_PARAMS_USER
```
default configuration is empty. You should put defines above (see other drivers implementations)

aabadie commented on drivers/soft_spi/soft_spi.c:
```diff
@@ -0,0 +1,202 @@
+#include "periph/spi.h"
+#include "periph/gpio.h"
+#include "soft_spi_params.h"
+#define ENABLE_DEBUG (1)
```
put 0

aabadie commented on tests/driver_soft_spi/soft_spi_params.h:
```diff
@@ -0,0 +1,45 @@
+/*
```
This file should be moved to drivers/soft_spi/include

plushvoxel replied:
This is thought to be the way to let the use/developer set the parameters for the soft SPI.

aabadie commented on tests/driver_soft_spi/soft_spi_params.h:
```diff
@@ -0,0 +1,45 @@
+ * @{
+ *
+ * @file
+ * @brief       Sotware SPI configuration
```
So**f**tware

aabadie commented on tests/driver_soft_spi/soft_spi_params.h:
```diff
@@ -0,0 +1,45 @@
+#endif
+
+/**
+ * @brief   Sotware SPI port descriptor array
```
Software

aabadie commented on tests/driver_soft_spi/Makefile:
```diff
@@ -0,0 +1,9 @@
+USEMODULE += soft_spi
+
+CFLAGS += -I$(CURDIR)
+#CFLAGS += -DSOFT_SPI_PARAMS_USER="{.mosi_pin = GPIO_PIN(0, 7), .miso_pin = GPIO_UNDEF, .clk_pin = GPIO_PIN(0, 6), .initialized = false,}"
```
This line can be removed

aabadie commented on tests/driver_soft_spi/Makefile:
```diff
@@ -0,0 +1,9 @@
+
+USEMODULE += soft_spi
+
+CFLAGS += -I$(CURDIR)
```
I don't think we need this.

plushvoxel replied:
The CFLAGS or the USEMODULE?

The CFLAGS ist necessary because else the compiler would disregard the soft_spi_params.h in the project directory.

aabadie commented on tests/driver_soft_spi/README.md:
```diff
@@ -0,0 +1,8 @@
+Expected result
```
Can you give more details about how to verify that the test works ?

PeterKietzmann commented on drivers/soft_spi/soft_spi.c:
```diff
@@ -0,0 +1,202 @@
+    if (!curr_soft_spi->initialized) {
+        spi_init_pins(bus);
+    }
+    spi_init_cs(bus, cs);
```
Why do you initialize pins and CS in each `spi_acquire`?

plushvoxel replied:
The programmer might use an already initialized software spi bus with a new CS pin.

PeterKietzmann commented:
@plushvoxel, @all, I PRed against this branch [see](https://github.com/plushvoxel/RIOT/pull/1). Please review them changes and merge if possible. Note: I still did not test other platforms which needs to be done. It seems there is a problem e.g. with samr21 boards

PeterKietzmann commented:
I created an other PR with small fixes (see [here](https://github.com/plushvoxel/RIOT/pull/2)). @plushvoxel did you ever test you implementation with an other board than stm32f4discovery? The good news, it generally works. The bad news is the interference with usual hardware SPI and the `periph_common ` implementation in e.g. with sam0_common. As you can see I added a `DISABLE_MODULE` to the applications make file. Additionally I applied this patch. Any ideas for an overall solution?
```
From 337fef3a8a981a38d96409a3d429ae4b019b3756 Mon Sep 17 00:00:00 2001
From: Peter Kietzmann <peter.kietzmann@haw-hamburg.de>
Date: Wed, 23 Aug 2017 08:50:52 +0200
Subject: [PATCH 1/1] Don't use HW SPI features

Signed-off-by: Peter Kietzmann <peter.kietzmann@haw-hamburg.de>
---
 cpu/sam0_common/include/periph_cpu_common.h | 2 +-
 drivers/periph_common/init.c                | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/cpu/sam0_common/include/periph_cpu_common.h b/cpu/sam0_common/include/periph_cpu_common.h
index 461b1e2..b9e1ff1 100644
--- a/cpu/sam0_common/include/periph_cpu_common.h
+++ b/cpu/sam0_common/include/periph_cpu_common.h
@@ -34,7 +34,7 @@ extern "C" {
  * @brief   Use shared SPI functions
  * @{
  */
-#define PERIPH_SPI_NEEDS_INIT_CS
+//#define PERIPH_SPI_NEEDS_INIT_CS
 #define PERIPH_SPI_NEEDS_TRANSFER_BYTE
 #define PERIPH_SPI_NEEDS_TRANSFER_REG
 #define PERIPH_SPI_NEEDS_TRANSFER_REGS
diff --git a/drivers/periph_common/init.c b/drivers/periph_common/init.c
index 3c13f15..d9fe2fc 100644
--- a/drivers/periph_common/init.c
+++ b/drivers/periph_common/init.c
@@ -25,7 +25,7 @@ void periph_init(void)
     /* initialize configured SPI devices */
 #ifdef SPI_NUMOF
     for (unsigned i = 0; i < SPI_NUMOF; i++) {
-        spi_init(SPI_DEV(i));
+        //spi_init(SPI_DEV(i));
     }
 #endif
 }
-- 
2.7.4
```

plushvoxel commented:
I only have some Arduino boards and Raspberry Pis at hand :/ Could test with those.

PeterKietzmann commented:
We have a lot of hardware in the office next to yours!

haukepetersen commented on drivers/soft_spi/soft_spi.c:
```diff
@@ -0,0 +1,207 @@
+/**
+ * @brief   Allocate one lock per SPI device
+ */
+static mutex_t locks[1];
```
do you allow for more than 1 soft bus at a time? If yes -> there should be no hard coded `1` here. If no -> no need for an array...

plushvoxel replied:
✓

haukepetersen commented on drivers/soft_spi/soft_spi.c:
```diff
@@ -0,0 +1,207 @@
+
+static inline bool spi_bus_is_valid(spi_t bus)
+{
+    unsigned int spi_num = (unsigned int) bus;
```
This won't always work. If some board would define the `spi_t` type to something else than an unsigned (e.g. a register base address), this cast here will fail... 

This is some ugly problem. The general solution would be to `#ifndef MODULE_SOFT_SPI` any redefinition of `spi_t` and `SPI_DEV`. But this would have to be done potentially for every board -> very ugly dependency and also this does not scale well... So I guess we need to think of something nicer.

haukepetersen replied:
Also, if `gpio_t` is not overridden by the CPU, there is no need for this cast, as `unsigned int` is the default for `spi_t`...

haukepetersen commented on drivers/soft_spi/include/soft_spi_params.h:
```diff
@@ -0,0 +1,59 @@
+        .miso_pin = SOFT_SPI_PARAM_MISO,
+        .mosi_pin = SOFT_SPI_PARAM_MOSI,
+        .clk_pin  = SOFT_SPI_PARAM_CLK,
+        .cs_pin   = SOFT_SPI_PARAM_CS,
```
Using a virtual hardware chip select line here is pretty much overkill. As a soft SPI has per definition no hardware chip select line, there is no reason trying to emulate one... I would strongly opt for dropping it.

plushvoxel replied:
✓

haukepetersen commented on drivers/soft_spi/soft_spi.c:
```diff
@@ -0,0 +1,207 @@
+    assert(soft_spi_config[bus].mosi_pin != soft_spi_config[bus].clk_pin);
+    assert(soft_spi_config[bus].miso_pin != soft_spi_config[bus].clk_pin);
+    /* mandatory pins */
+    assert((GPIO_UNDEF != soft_spi_config[bus].mosi_pin) || (GPIO_UNDEF != soft_spi_config[bus].miso_pin));
```
be consitent: either check all three in a single assert or use 3 asserts...

haukepetersen replied:
ups, my bad, didn't notice the OR, never mind.

haukepetersen commented on drivers/soft_spi/soft_spi.c:
```diff
@@ -0,0 +1,207 @@
+    /* initialize clock pin */
+    gpio_init(soft_spi_config[bus].clk_pin, GPIO_OUT);
+    /* initialize optional pins */
+    if (GPIO_UNDEF != soft_spi_config[bus].mosi_pin) {
```
the assert above already make sure that `mosi_pin` can not be `GPIO_UNDEF`, so this if condition here is redundant.

haukepetersen replied:
see above, never mind.

haukepetersen commented on drivers/soft_spi/soft_spi.c:
```diff
@@ -0,0 +1,207 @@
+    if (GPIO_UNDEF != soft_spi_config[bus].mosi_pin) {
+        gpio_init(soft_spi_config[bus].mosi_pin, GPIO_OUT);
+        gpio_clear(soft_spi_config[bus].mosi_pin);
+    }
```
pin initialization for `miso_pin` missing?!

plushvoxel replied:
✓

haukepetersen commented on drivers/soft_spi/soft_spi.c:
```diff
@@ -0,0 +1,207 @@
+int spi_acquire(spi_t bus, spi_cs_t cs, spi_mode_t mode, spi_clk_t clk)
+{
+    /* Right now we don't care about speed */
+    (void)clk;
```
this should be also stated in the module description...

plushvoxel replied:
Do you mean in the header file in 
```
/**
 * @defgroup    driver_soft_spi Soft SPI
 * @ingroup     drivers
 * @brief
```
?

haukepetersen commented on drivers/soft_spi/soft_spi.c:
```diff
@@ -0,0 +1,207 @@
+    /* lock bus */
+    mutex_lock(&locks[bus]);
+
+    if (mode < SPI_MODE_0 || SPI_MODE_3 < mode) {
```
same here: what happens if a CPU has redefined the SPI_MODE_x values to something else?! This would break this code here.

plushvoxel replied:
✓

haukepetersen commented on drivers/soft_spi/soft_spi.c:
```diff
@@ -0,0 +1,207 @@
+    mutex_unlock(&locks[bus]);
+}
+
+uint8_t _transfer_one_byte(spi_t bus, spi_cs_t cs, bool cont, uint8_t out)
```
`static` missing. Better: use `static inline` for this function.

plushvoxel replied:
✓

haukepetersen commented on drivers/soft_spi/soft_spi.c:
```diff
@@ -0,0 +1,207 @@
+
+uint8_t _transfer_one_byte(spi_t bus, spi_cs_t cs, bool cont, uint8_t out)
+{
+    assert(spi_bus_is_valid(bus));
```
move this `assert` into `spi_transfer_bytes()` -> no need to check the `bus` parameter for each single byte...

plushvoxel replied:
✓

haukepetersen commented on drivers/soft_spi/soft_spi.c:
```diff
@@ -0,0 +1,207 @@
+            gpio_set(soft_spi_config[bus].cs_pin);
+        }
+    }
+}
```
newline at EOF missing

plushvoxel replied:
✓

haukepetersen commented on drivers/soft_spi/soft_spi.c:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    for (size_t i = 0; i < len; i++) {
+        if (NULL != out) {
```
I think it would make sense to change this block to the following:
```c
uint8_t tmp = (out) ? out[i] : 0;
retval = _transfer_one_byte(bus, cs, true, &tmp);
```
This would also make the inlining more efficient I think.

plushvoxel replied:
✓

haukepetersen commented on drivers/soft_spi/soft_spi.c:
```diff
@@ -0,0 +1,207 @@
+    mutex_unlock(&locks[bus]);
+}
+
+uint8_t _transfer_one_byte(spi_t bus, spi_cs_t cs, bool cont, uint8_t out)
```
also: both `cs` and `cont` parameters are unused in this function, right?

haukepetersen commented on tests/driver_soft_spi/Makefile:
```diff
@@ -0,0 +1,17 @@
+
+BOARD ?= stm32f4discovery
+
+BOARD_WHITELIST := stm32f4discovery samr21-xpro
```
Why only these two boards? Shouldn't the `soft_spi` run for every board we support?

plushvoxel replied:
In theory yes, but it hasn't been tested on many devices yet.

haukepetersen commented on tests/driver_soft_spi/Makefile:
```diff
@@ -0,0 +1,17 @@
+BOARD_WHITELIST := stm32f4discovery samr21-xpro
+
+USEMODULE += soft_spi
+USEMODULE += periph_common
```
better: add `periph_common` as dependency to `soft_spi` in `drivers/Makefile.dep`!

plushvoxel replied:
✓

haukepetersen commented:
In general I like this PR very much, was thinking of implementing something like this for a long time...

But this PR shows again the problem with our periph interface: it was not designed to multiplex different implementations of its API. So the main problem here is, that we have no control over CPUs redefining the semantics of the default SPI specific types and defines (e.g. `spi_t`, `SPI_DEV`, ...).

Not sure how to proceed here, ideas anyone?

plushvoxel commented on drivers/include/soft_spi.h:
```diff
@@ -0,0 +1,57 @@
+ * @brief Software SPI port descriptor
+ */
+typedef struct {
+    gpio_t miso_pin;              /**<MOSI pin */
```
in every line or just that one?

plushvoxel commented on drivers/soft_spi/soft_spi.c:
```diff
@@ -0,0 +1,207 @@
+    assert(soft_spi_config[bus].mosi_pin != soft_spi_config[bus].clk_pin);
+    assert(soft_spi_config[bus].miso_pin != soft_spi_config[bus].clk_pin);
+    /* mandatory pins */
+    assert((GPIO_UNDEF != soft_spi_config[bus].mosi_pin) || (GPIO_UNDEF != soft_spi_config[bus].miso_pin));
```
The assertion in line 71 is the only one that need a boolean evaluation.

plushvoxel commented on drivers/soft_spi/soft_spi.c:
```diff
@@ -0,0 +1,207 @@
+    /* initialize clock pin */
+    gpio_init(soft_spi_config[bus].clk_pin, GPIO_OUT);
+    /* initialize optional pins */
+    if (GPIO_UNDEF != soft_spi_config[bus].mosi_pin) {
```
The assertion in line 71 make sure that `mosi_pin` and `miso_pin` can't both be `GPIO_UNDEF` at the same time.

haukepetersen commented on drivers/include/soft_spi.h:
```diff
@@ -0,0 +1,57 @@
+ * @brief Software SPI port descriptor
+ */
+typedef struct {
+    gpio_t miso_pin;              /**<MOSI pin */
```
this line and the lines below as far as I can see

plushvoxel replied:
✓

plushvoxel commented on drivers/include/soft_spi.h:
```diff
@@ -0,0 +1,57 @@
+    spi_mode_t spi_mode;          /**<data and clock polarity */
+} soft_spi_t;
+
+
```
✓

plushvoxel commented on drivers/soft_spi/include/soft_spi_params.h:
```diff
@@ -0,0 +1,59 @@
+#define SOFT_SPI_PARAM_MISO        (GPIO_UNDEF)
+#endif
+#ifndef SOFT_SPI_PARAM_MOSI
+#define SOFT_SPI_PARAM_MOSI        (GPIO_PIN(0, 7))
```
✓

plushvoxel commented on drivers/soft_spi/include/soft_spi_params.h:
```diff
@@ -0,0 +1,59 @@
+        .cs_pin   = SOFT_SPI_PARAM_CS,
+        .spi_mode = SPI_MODE_0
+    },
+};
```
Would I do so by also including
```
#ifndef SOFT_SPI_MODE
#define SOFT_SPI_MODE (SPI_MODE_0)
#endif
```
?

plushvoxel replied:
✓

plushvoxel commented:
Ahhh, I used the wrong local account settings... meh

Will squash as soon as I tested it on hardware

plushvoxel commented:
I now made a medium big copy-pasta of the spi.h; removed things like hardware CS pin stuff and search and replaced "spi" with "soft_spi".
This results in software spi being independent from the hardware spi api but shares functions with similar signature and a project can be moved from using hard spi to soft spi and vice versa by search and replace.

PeterKietzmann commented on drivers/Makefile.dep:
```diff
@@ -184,6 +184,11 @@ ifneq (,$(filter sdcard_spi,$(USEMODULE)))
 
+ifneq (,$(filter soft_spi,$(USEMODULE)))
+  FEATURES_REQUIRED += periph_gpio
+  USEMODULE += periph_common
```
I think this is not needed anymore

plushvoxel replied:
:heavy_check_mark:

PeterKietzmann commented on drivers/include/soft_spi.h:
```diff
@@ -0,0 +1,286 @@
+ * @brief   Chip select pin type overlaps with gpio_t so it can be casted to
+ *          this
+ */
+#ifndef HAVE_SOFT_SPI_CS_T
```
I guess this will never be defined

plushvoxel replied:
:heavy_check_mark:

PeterKietzmann commented on drivers/include/soft_spi.h:
```diff
@@ -0,0 +1,286 @@
+ * - MODE_3: CPOL=1, CPHA=1 - The first data bit is sampled by the receiver on
+ *           the second falling SCK edge.
+ */
+#ifndef HAVE_SOFT_SPI_MODE_T
```
Ok what is the intention behind these checks? Is it just copy paste from spi.h?

plushvoxel replied:
jup, copy paste error
:heavy_check_mark:

PeterKietzmann commented on drivers/include/soft_spi.h:
```diff
@@ -0,0 +1,286 @@
+    SOFT_SPI_CLK_1MHZ,           /**< drive the SPI bus with 1MHz */
+    SOFT_SPI_CLK_5MHZ,           /**< drive the SPI bus with 5MHz */
+    SOFT_SPI_CLK_10MHZ           /**< drive the SPI bus with 10MHz */
+} soft_spi_clk_t;
```
I think we agreed on running as fast as possible, why did you introduce speed types?

plushvoxel replied:
to have similar signature with the function of hardware spi, the function still have the clock speed parameter.
also if programmer do a search and replace to adapt the hardware spi using code to use soft spi, the don't have to manually change the clock speed value/constants.
speed may also be implemented later in time when needed although right now the parameter is ignored in the transmission

PeterKietzmann commented on drivers/include/soft_spi.h:
```diff
@@ -0,0 +1,286 @@
+ * This function is intended to be called by the board initialization code
+ * during system startup to prepare the (shared) SPI device for further usage.
+ * It uses the board specific initialization parameters as defined in the
+ * board's `periph_conf.h`.
```
This is not correct, is it?

plushvoxel replied:
:heavy_check_mark:

PeterKietzmann commented on drivers/include/soft_spi.h:
```diff
@@ -0,0 +1,286 @@
+ * used. But there are some devices (e.g. CC110x), that use SPI bus lines also
+ * for other purposes and need the option to dynamically re-configure one or
+ * more of the used pins. So they can take control over certain pins and return
+ * control back to the SPI driver using this function.
```
IMO it is not needed to keep the whole copy-pasta text here. Just state what the function does and that pins should be initialized after `spi_init`

plushvoxel replied:
copy paste error again
:heavy_check_mark:

PeterKietzmann commented on drivers/include/soft_spi.h:
```diff
@@ -0,0 +1,286 @@
+ * @brief   Initialize the given chip select pin
+ *
+ * The chip select can be any generic GPIO pin (e.g. GPIO_PIN(x,y)), or it can
+ * be a hardware chip select line. The existence and number of hardware chip
```
How can it be **hardware** CS?

plushvoxel replied:
:heavy_check_mark:

PeterKietzmann commented on drivers/include/soft_spi.h:
```diff
@@ -0,0 +1,286 @@
+#endif
+
+#endif /* SOFT_SOFT_SPI_H */
+/** @} */
```
Please check the whole documentation of this file. It can be drastically reduced. One thing that I missed is the hint about ignored clock speeds in the "module description" at the beginning of this file. Remember that you want to explain the soft SPI driver here, not any generic or hardware dependent interface

plushvoxel replied:
:heavy_check_mark:

PeterKietzmann commented on drivers/soft_spi/soft_spi.c:
```diff
@@ -0,0 +1,209 @@
+void soft_spi_init(soft_spi_t bus)
+{
+    DEBUG("Soft SPI init\n");
+    /* This is left empty because the periph_init function will call this for
```
What is left empty and what's the relation the the hardware? Please make clear or remove.

plushvoxel replied:
that was a leftover comment
:heavy_check_mark:

PeterKietzmann commented on tests/driver_soft_spi/Makefile:
```diff
@@ -0,0 +1,14 @@
+
+USEMODULE += soft_spi
+
+CFLAGS += "-D SOFT_SPI_PARAMS_CUSTOM = {\
```
This looks weird? `SOFT_SPI_PARAMS_CUSTOM` only sets the miso pin here? I would say either remove the CFLAG and rely on the params file or add a complete entry. First solution should be easiest.

plushvoxel replied:
```
CFLAGS += "-D SOFT_SPI_PARAMS_CUSTOM = {\
                                           .miso_pin = GPIO_UNDEF,\
																				   .mosi_pin = (GPIO_PIN(0, 0)),\
																				   .clk_pin  = (GPIO_PIN(0, 1)),\
																		   }"
```
sets all three pins

PeterKietzmann commented on tests/driver_soft_spi/Makefile:
```diff
@@ -0,0 +1,14 @@
+
+CFLAGS += "-D SOFT_SPI_PARAMS_CUSTOM = {\
+                                           .miso_pin = GPIO_UNDEF,\
+																				   .mosi_pin = (GPIO_PIN(0, 0)),\
```
Remove all these blank lines

plushvoxel replied:
ah, you may have bigger tab space than i do

PeterKietzmann commented on tests/driver_soft_spi/main.c:
```diff
@@ -0,0 +1,72 @@
+    soft_spi_cs_t cs = GPIO_PIN(0, 2);
+
+    /* Initialize software SPI device. This function is usually
+       called during board initialization*/
```
Not anymore. Please adapt

haukepetersen commented on drivers/soft_spi/Makefile:
```diff
@@ -0,0 +1,3 @@
+MODULE = soft_spi
```
this line is not needed, the module name is deducted from the path automatically...

haukepetersen commented on tests/driver_soft_spi/Makefile:
```diff
@@ -0,0 +1,10 @@
+
+USEMODULE += soft_spi
+
+CFLAGS += "-D SOFT_SPI_PARAMS_CUSTOM={.miso_pin=GPIO_UNDEF, .mosi_pin=(GPIO_PIN(0, 4)), .clk_pin=(GPIO_PIN(0, 6)),}"
```
drop this line, as it is specific for your test env.

haukepetersen commented on tests/driver_soft_spi/Makefile:
```diff
@@ -0,0 +1,10 @@
+APPLICATION = driver_soft_spi
+include ../Makefile.tests_common
+
+BOARD ?= stm32f4discovery
```
I would opt for leaving native as default here.

haukepetersen commented on tests/driver_soft_spi/README.md:
```diff
@@ -0,0 +1,10 @@
+
+Notes
+==========
+- The test disables all pre-initialized hardware SPI devices
```
nope, at least not in the code below...

haukepetersen commented on tests/driver_soft_spi/README.md:
```diff
@@ -0,0 +1,10 @@
+==========
+- The test disables all pre-initialized hardware SPI devices
+- The test does not cover MISO inputs
+- The test ignores clock speeds
```
So if we ignore clock speeds and doe not cover MISO, there is not much functionality of a SPI bus left, is there?

haukepetersen commented:
Code looks ok to me, but I think the test application can/shoud be significantly improved. We could at least tell the user to connect MISO and MOSI pins via jumper wires and write the test so that it compares incoming and outgoing data.

Also I think the clock configuration must be implement to make the `soft_spi` driver meaningful. At least the low clock speeds can easily be implemented using `xtimer`, and we could simply restrict the clock speed to 100 and 400khz.

plushvoxel commented:
Okay,
the thing is that @PeterKietzmann and I settled on implementing just the transmitting part for now. But if the consensus is that the receiving part should be implemented to for the first PR, then this should be possible too. There the test requires a logic analyzer.

The clock speed part I also had implemented at the beginning and discarded in the end, because it didn't perform well. Most SPI slaved can handle up to a couple MHz, which won't be exceeded by using the high level pin operations with no wait function in between.

haukepetersen commented:
> Most SPI slaved can handle up to a couple MHz

Ture, but `most` is not good enough here, we need a precise, defined, and deterministic state...

PeterKietzmann commented:
@haukepetersen offline we agreed on 
- Adapting documentation
- Introducing speed limitation (again)
- Say with MOSI implementation

@plushvoxel when will you provide this? I would like to see this as part of the release

PeterKietzmann commented:
@haukepetersen ping! This is a candidate to be merged before feature freeze

haukepetersen commented on drivers/Makefile.dep:
```diff
@@ -193,6 +193,11 @@ ifneq (,$(filter sdcard_spi,$(USEMODULE)))
   USEMODULE += xtimer
 endif
 
+ifneq (,$(filter soft_spi,$(USEMODULE)))
```
>  driver dependencies (in alphabetical order)

`sht11` vs `soft_spi` vs. `si70xx`...

haukepetersen commented:
took the freedom to change the PR title..

gebart commented:
Why is the CS pin being asserted once and released again before the real transmission starts in the logic analyzer graphs above?

PeterKietzmann commented:
Thanks for your feedback! @gebart that is legacy issues. The screen-shot shows the first four bytes which are transmitted in all four modes (thus, can not be interpreted with one analyzer setting). 

![softspi_modes](https://user-images.githubusercontent.com/7765855/31546175-7fabd2e6-b022-11e7-97d1-a830d250bed7.png)

The `.logicdata` can be found [here](https://owncloud.inet.haw-hamburg.de/index.php/s/1VNjKhjLkPATFqY)

PeterKietzmann commented:
@haukepetersen @kaspar030 please remove your change requests

PeterKietzmann commented:
@plushvoxel please have a look at [Murdocks output](https://ci.riot-os.org/RIOT-OS/RIOT/7156/c707b684248a0fb2bc84089d42ebd52be6fba2e9/output.html) (search for "error" in there). Some minor fixes are reqired before we can finally merge that thing :-)

PeterKietzmann commented:
All lights are green! Thanks @plushvoxel. ACK and go
